### ISSUE TITLE
namespace declarations produces "C++ Parser can't read code. Declaration is skipped." errors.

### ISSUE THREAD

**Issue by phcouton** [SonarOpenCommunity/sonar-cxx#2037]
**This problem occurs only with the current code in the GitHub repository (2.0.0).
It does not occur with the latest published version (1.3.3) !**

### Description

namespaces declarations such as `namespace foo { /* namespace contents */ }` produces "C++ Parser can't read code. Declaration is skipped." errors in SonarQube. Such errors occurs on code that compiles without problem.

There is the following message in the output during the analysis:
> Source code parser: 1 syntax error(s) detected. Syntax errors could cause invalid software metric values. Root cause are typically missing includes, missing macros or compiler specific extensions.

### Steps to reproduce the problem

1. Create a project with a C++ file containing this:
```C++
namespace foo {
}
```
2. Analyze it

[Sample project](https://github.com/SonarOpenCommunity/sonar-cxx/files/5835408/namespaceParseErrorProject.zip)

### Expected behavior

Analysis without error.

### Actual behavior

Unexpected parsing errors.

### LOG file

[Resulting log](https://github.com/SonarOpenCommunity/sonar-cxx/files/5835446/sonar.log). See line 309 for the error.

### Related information

* cxx plugin version: 2.0.0-SNAPSHOT, clone of the GitHub repository on January 15, branch master
  problem not seen with version 1.3.3
* SonarQube version7.9.1
* Sonar scanner version 4.4.0

**Comment by guwirth** [comment]
Hi @phcouton,

thanks for your feedback.

I tried it with the SSLR toolkit, there it is working fine?
![grafik](https://user-images.githubusercontent.com/6077367/105037312-145d0c00-5a5e-11eb-98f0-e9d92d29e11d.png)

One thing I saw is that you are testing with SQ 7.9: cxx plugin 2.0 will be released for SQ 8.7 LTS and Java 11 (currently tested with SQ 8.6). Need further investigation...

Regards,

**Comment by guwirth** [comment]
Hi @phcouton,

there is only this generic message in the LOG file:
```
11:13:49.669 WARN: Source code parser: 1 syntax error(s) detected. Syntax errors could cause invalid software metric values. Root cause are typically missing includes, missing macros or compiler specific extensions.
```
Why do you think it is a namespace error?

Regards,

**Comment by phcouton** [comment]
> 
> 
> Hi @phcouton,
> 
> thanks for your feedback.
> 
> I tried it with the SSLR toolkit, there it is working fine?
> ![grafik](https://user-images.githubusercontent.com/6077367/105037312-145d0c00-5a5e-11eb-98f0-e9d92d29e11d.png)
> 
> One thing I saw is that you are testing with SQ 7.9: cxx plugin 2.0 will be released for SQ 8.7 LTS and Java 11 (currently tested with SQ 8.6). Need further investigation...
> 
> Regards,

I confirm the same for the SSLR toolkit use.

For tests, I can only access a SonarQube server 7.9 for now. To anticipate the next LTS coming soon, we are deploying a 8.6 server; I will remake my tests with it when it will be ready.

**Comment by phcouton** [comment]
> 
> 
> Hi @phcouton,
> 
> there is only this generic message in the LOG file:
> 
> ```
> 11:13:49.669 WARN: Source code parser: 1 syntax error(s) detected. Syntax errors could cause invalid software metric values. Root cause are typically missing includes, missing macros or compiler specific extensions.
> ```
> 
> Why do you think it is a namespace error?
> 
> Regards,

1) because I made a test where the whole project was this single declaration (see the sample projet joined)
2) Because of that:
![image](https://user-images.githubusercontent.com/13865666/105046933-fc8b8500-5a69-11eb-9780-32d53fd125f0.png)

**Comment by guwirth** [comment]
> Why do you think it is a namespace error?
>> Because of that:
![image](https://user-images.githubusercontent.com/13865666/105046933-fc8b8500-5a69-11eb-9780-32d53fd125f0.png)

Thx, will have a look.

**Comment by guwirth** [comment]
@phcouton should be fixed, was a problem with error recovery (#2038).
Please have also a look to the draft https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Upgrade-the-Plugin

**Comment by phcouton** [comment]
> 
> 
> @phcouton should be fixed, was a problem with error recovery (#2038).
> Please have also a look to the draft https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Upgrade-the-Plugin

Got the fix and tested it, it works as expected with SQ 8.6 (and even 7.9), tanks :slightly_smiling_face:

For the wiki draft, it seems OK to me. The only point that should be clarified is the configuration property to be used for the files suffixes: the difference between `sonar.cxx.api.file.suffixes` and `sonar.cxx.file.suffixes` is not clear to me. On my side, I made my tests using `sonar.cxx.file.suffixes` and it works well.

On the other hand, due to the language key change, I lost my quality profile on the first server start with the new plugin. I found a tip to avoid that:

1. backup your old quality profile before using plugin V2.0.0;
2. with a text editor, change `<language>c++</language>` into `<language>cxx</language>` in the backup file (only one occurence at the begining of the file);
3. install the plugin V2.0.0 and restart the SQ server;
4. restore the modified quality profile.

Maybe it's worth documenting this in the wiki?

**Comment by guwirth** [comment]
@phcouton thanks for your feeback, I will improve the documentation:

* `sonar.cxx.api.file.suffixes` is the setting for the undocumented API check/metrics: https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-checks/src/main/java/org/sonar/cxx/checks/api/UndocumentedApiCheck.java
* Quality Profile: I will add your hint.
* I will have a look if final SQ 8.7 LTS is still compatible with SQ 7.9 LTS. If yes we will release it for both versions.

**Comment by guwirth** [comment]
@phcouton would `sonar.cxx.metric.api.file.suffixes` be better?

**Comment by phcouton** [comment]
> 
> 
> @phcouton would `sonar.cxx.metric.api.file.suffixes` be better?

@guwirth Yes, I think. `sonar.cxx.api.file.suffixes` and `sonar.cxx.file.suffixes` are too close, I find this error-prone.

**Comment by guwirth** [comment]
@phcouton in latest 2.0 snapshot configuration parameter are renamed.
